### ISSUE TITLE
move through friendlies

### ISSUE THREAD

**Issue by Grokmoo** [Grokmoo/sulis#158]
Characters (party members and hostiles) should be able to move through friendly entities in combat mode.  they should not be able to end their movement on top of a friendly.  in particular, if they only have enough movement to move part of the path to their destination.

**Comment by picobyte** [comment]
Pathfinding occurs in sulis_module/src/area/path_finder.rs PathFinder::find() it seems. Since a unit is not supposed to end on a friendly unit, It's easiest if it's not possible to choose a target location occupied by a friendly unit. This seems currently the case. Secondly, if all remaining movements for a turn end on friendly units, then the remaining moves are impassible.
To enable testing whether a unit is a friendly requires more information during pathfinding and in the passability check.

**Comment by Grokmoo** [comment]
we cannot compute ahead of time where a mover will run out of AP.  Scripts could modify movement rate at any time, and even without that creatures could enter or exit auras / surfaces that affect movement rate.

We already pass a list of entities to ignore in passability checks to the pathfinder.  We could easily enough include all friendlies in this list.  But, then we need to handle the case when a unit runs out of AP on top of another.  My preferred approach is to move back along the path they already took until an open spot is found.

An alternative would be to just bump them out to a nearby valid spot.  There is already logic that accommodates this in party_bump_handler.  But, this seems like the player could exploit this behavior to "cheat" a bit.  Conversely, it may be unpredictable where an entity ends up to someone without a full understanding of the rules.

There is a potential performance pitfall in making the entities_to_ignore list large.  Will need to benchmark.

**Comment by picobyte** [comment]
Let's brainstorm about the preferred option (walk back) first.
We may be able to exclude some friendlies from the entities_to_ignore list if they do not occur in the square between entity and dest.
The `checker.passable()` in sulis_module/src/area/path_finder.rs `PathFinder::find()` currently returns
true/false but may require a third state to enable a walk back if we run out of moves.

Off topic but the get_neighbors() seems to return all 4 neigbours, but I think we could skip neighbours if they were our last or fore last position (if we already had the steps). I think this is
what self.closed contains, but isn't just testing these 2 sufficient?

**Comment by picobyte** [comment]
I've made an attempt to resolve this in branch:
https://github.com/picobyte/sulis/tree/move_through_friendlies
It's currently only compile tested. I hope it's on the right track.

**Comment by Grokmoo** [comment]
This is a bit tricky.  Unfortunately I think there are some issues with what is here.  There are essentially two different problems we are dealing with:
  1. We don't want the actual goal of the path find to end up in an invalid position.  I think the appropriate way to handle this is to modify the is_goal function and require the checked point to not overlap any entities (but only when combat is active.  when not in combat, it is ok for entities to end up overlapping).
  2. If a creature is mid move and runs out of AP before completing the path, we don't want them to end up in an invalid position.  This needs to happen after the path_find, probably in the move_animation.  Basically, when ending / canceling the move animation, we need to check if the mover is in an invalid position.  If they are, walk back along the path until a valid position is found and place them there.

On changing entities_to_ignore to HashSet - I've done quite a lot of benchmarking and optimizing here around the data structures and arguments being passed around.  To change this, I would want to see some actual timing numbers in a few different cases:
  * entities_to_ignore size is 3 - this is what it will be for a player with a full party.  HashSet should be slower.
  * entities_to_ignore size is 40ish - this is what it might be for an AI opponent with a lot of other opponents in the current area.  Will HashSet be faster?
  * entities_to_ignore size 10ish - a smaller number of AI opponents if we are more intelligent, such as limiting to fairly close-by entities.

On get_neighbors() - yes, checking the closed set handles this case.  This could maybe be optimized?  I would again want to see firm timing numbers before making any changes.

**Comment by picobyte** [comment]
Thanks again for your remarks. I will change back the HashSet to Vec, I'm not entirely sure yet how to benchmark. Also, not required this time, but if concerning a certain section in the code, could you add comments to the relevant line in the diff? Just mention here that you added comments and I will go through them.

Concerning the get_neigbors(), recurrances with in `closed`, which were before the before-last iteration, and not an impassible, indicate that the iteration can be skipped, I think. That must be a path that cannot be the best. I will try a hashmap and test whether this occurs.

The remainding remarks I will have to ponder.

**Comment by Grokmoo** [comment]
Hi, I appreciate your desire to help and effort on this.  But as I said before - I'm not going to merge optimizations without benchmarks that show a real improvement.

**Comment by picobyte** [comment]
That is why I put the optimization in a separate feature branch.

**Comment by picobyte** [comment]
Once the path is complete in the [`path_find()`](https://github.com/picobyte/sulis/blob/3d11b9fe7caa10a4a7bbf6615279656ba36db580/sulis_module/src/area/path_finder.rs#L220-L231) function I try to eliminate the ends of a path if they end on a party-members' space:
```
                let mut path = self.reconstruct_path(current);

                loop {
                    if path.len() == 1 && path[0].x == start_x && path[0].y == start_y {
                        debug!("Found path with no moves.");
                        return None;
                    }
                    if !checker.in_friend_space(path.last()) {
                        break;
                    }
                    path.pop();
                }
```
..but somehow this [`in_friend_space()`](https://github.com/picobyte/sulis/blob/3d11b9fe7caa10a4a7bbf6615279656ba36db580/sulis_state/src/path_finder.rs#L79) function doesn't seem to work, could it be that the entity_grid, in the code below, lists the entities of the adjacent spot?
```
    fn in_friend_space(&self, op: Option<&Point>) -> bool {
        if let Some(&p) = op {
            let index = (p.x + p.y * self.width) as usize;
            for i in self.entity_grid[index].iter() {
                if self.entities_to_ignore.contains(i) {
                    return true;
                }
            }
        }
        false
    }
}
```

**Comment by Grokmoo** [comment]
> 
> 
> Once the path is complete in the [`path_find()`](https://github.com/picobyte/sulis/blob/3d11b9fe7caa10a4a7bbf6615279656ba36db580/sulis_module/src/area/path_finder.rs#L220-L231) function I try to eliminate the ends of a path if they end on a party-members' space:
> 
> ```
>                 let mut path = self.reconstruct_path(current);
> 
>                 loop {
>                     if path.len() == 1 && path[0].x == start_x && path[0].y == start_y {
>                         debug!("Found path with no moves.");
>                         return None;
>                     }
>                     if !checker.in_friend_space(path.last()) {
>                         break;
>                     }
>                     path.pop();
>                 }
> ```

I don't think this is the right approach.  In this situation, you have already found a path (ignoring friendlies) and now are removing points at the end of the path until you reach an open space.  This may not be the optimal path.  Consider the case where the mover could move around a slightly longer way to reach a true open spot, rather than the computed way which leads to an invalid spot.  In this case, the method here will find a sub-optimal path.

In addition, we currently have a (perhaps not explicitly stated enough!) assumption that the pathfinder either finds a valid path within the specified destination constraints, or returns None.  Here, by moving back until a valid location is found, the mover may (in fact, most likely will!) end up outside the maximum distance bounds specified by the Destination.

I *believe* an approach that will work would be to modify the `is_goal` method in `sulis_module/src/area/path_finder.rs` to return false if a point is invalid due to friendly overlap.  My concern here is that this makes the `goal` function incompatible with the `dist_squared` Euclidean distance heuristic we are using.  I don't think this will cause failure of the algorithm though.

> 
> ..but somehow this [`in_friend_space()`](https://github.com/picobyte/sulis/blob/3d11b9fe7caa10a4a7bbf6615279656ba36db580/sulis_state/src/path_finder.rs#L79) function doesn't seem to work; I can still end my path on a friendly unit. Could it be that the entity_grid, in the code below, lists the entities of the adjacent spot, or that the path is changed after all this?
> 
> ```
>     fn in_friend_space(&self, op: Option<&Point>) -> bool {
>         if let Some(&p) = op {
>             let index = (p.x + p.y * self.width) as usize;
>             for i in self.entity_grid[index].iter() {
>                 if self.entities_to_ignore.contains(i) {
>                     return true;
>                 }
>             }
>         }
>         false
>     }
> ```

Please re-read my reply here: https://github.com/Grokmoo/sulis/issues/158#issuecomment-588600036 - especially item (2).

The issue you are (probably) experiencing is that a mover can run out of AP long after the move has already been calculated, and this cannot in general be predicted ahead of time.  This is the case where we may need to backtrack or otherwise find a valid position to bump the mover to.

To clarify, here is the sequence of events:

* AI or player request a move.  Path is calculated (entity's current AP is known, so we can estimate how far they can move).
* The calculated path is stored in `sulis_state/src/animation/move_animation.rs`.
* Once every specified number of milliseconds (based on the game's configured base animation speed), the mover is moved one square of this path.
* The movement can be interrupted, the rate of use of AP for the mover could change due to a surface, they could step on a trap and become unable to move at all, etc.  In any of these events, the mover could possibly end up standing in an intermediate square of the path, which could be invalid.  This is regardless of the validity of the computed path as a whole.

Apologies for being long winded; I hope that helps.

**Comment by picobyte** [comment]
Thank you again for your pointers, the move_animation and beyond seems what I missed, and I will have to look into that. Concerning the path_find, the is_goal function can be edited the simple way: 
```
    fn is_goal<T: LocationChecker>(&self, checker: &T, current: i32, dest_dist_squared: i32) -> bool {
        self.dist_squared(current) <= dest_dist_squared && !checker.in_friend_space(current)
    }
```
Also possible is below function, which could yield a more better solution, Euclidian distance-wise, but I think it may not be worth it. That may give less performance, and we do still need to edit the path if it ended on a friend space.
```
    fn is_goal<T: LocationChecker>(&self, checker: &T, mut current: i32, dest_dist_squared: i32) -> bool {
        while self.dist_squared(current) <= dest_dist_squared {
            if !checker.in_friend_space(current) {
                return true;
            }
            current = match self.came_from.get(&current) {
                None => break,
                Some(point) => *point,
            };
        }
        false
    }
```

**Comment by picobyte** [comment]
I find the logic in Anim::update() confusing.
```
        let keep = (self.duration_millis.less_than(elapsed) && self.ok_to_remove())
            || self.marked_for_removal.get();
        !keep
```
should that maybe be `|| !self.ok_to_remove())` instead?

**Comment by Grokmoo** [comment]
> 
> 
> Thank you again for your pointers, the move_animation and beyond seems what I missed, and I will have to look into that. Concerning the path_find, the is_goal function can be edited the simple way:
> 
> ```
>     fn is_goal<T: LocationChecker>(&self, checker: &T, current: i32, dest_dist_squared: i32) -> bool {
>         self.dist_squared(current) <= dest_dist_squared && !checker.in_friend_space(current)
>     }
> ```

Yes, this looks correct.

> Also possible is below function, which could yield a more better solution, Euclidian distance-wise, but I think it may not be worth it. That may give less performance, and we do still need to edit the path if it ended on a friend space.
> 
> ```
>     fn is_goal<T: LocationChecker>(&self, checker: &T, mut current: i32, dest_dist_squared: i32) -> bool {
>         while self.dist_squared(current) <= dest_dist_squared {
>             if !checker.in_friend_space(current) {
>                 return true;
>             }
>             current = match self.came_from.get(&current) {
>                 None => break,
>                 Some(point) => *point,
>             };
>         }
>         false
>     }
> ```

No, this is not necessary (and in fact would cause problems).  The path find will not end on a point where `is_goal` returns false.  You do not need to edit the path because it cannot end on a "friend space" as long as `is_goal` returns false for all friend spaces.

> 
> 
> I find the logic in Anim::update() in `sulis_state/src/animation.rs` confusing.
> 
> ```
>         let keep = (self.duration_millis.less_than(elapsed) && self.ok_to_remove())
>             || self.marked_for_removal.get();
>         !keep
> ```
> 
> should keep be named remove?

Yeah, that seems right.

**Comment by picobyte** [comment]
If I understand correctly, the move_animation needs to be adapted. To this extend, in [`move_towards_point()`](https://github.com/Grokmoo/sulis/blob/17b7b01e89651420f7c3651bc52b6ee2dde72815/sulis_state/src/path_finder.rs#L83) we could copy the entities_to_ignore in the MoveAnimModel struct (wasn't able to store it as a reference, because of life time problems).
```
let mut anim = animation::move_animation::new(entity, path,
                entities_to_ignore.to_vec(),
                Config::animation_base_time_millis());
```
This enables the move_animation update() function to check whether a position in the animation is occupied by a friendly unit.
```
    let p = model.path[frame_index];
    let area_state = GameState::get_area_state(&model.area_id).unwrap();
    let mut area_state = area_state.borrow_mut();

    if !area_state.move_entity(mover, p.x, p.y, move_ap as u32) {
        marked_for_removal.set(true);
        return;
    }

    let index = (p.x + p.y * area_state.area.width) as usize;

    if area_state.entity_grid[index].iter().any(|i| model.entities_to_ignore.contains(i)) {
        // do something here
    } else {
        // or here
    }
```
Either we should
* try predict that the remaining path is all on friend space, and refuse to continue.
* do the walk on friend space, but store a walk_back for each on friend space site, and add the walk back once we ended on a friend space.

**Comment by Grokmoo** [comment]
> 
> 
> If I understand correctly, the move_animation needs to be adapted something like this: in [`move_towards_point()`](https://github.com/Grokmoo/sulis/blob/17b7b01e89651420f7c3651bc52b6ee2dde72815/sulis_state/src/path_finder.rs#L83) we could copy the `entities_to_ignore` in the [`MoveAnimModel`](https://github.com/Grokmoo/sulis/blob/17b7b01e89651420f7c3651bc52b6ee2dde72815/sulis_state/src/animation/move_animation.rs#L179) struct (wasn't able to store it as a reference, because of life time problems).

We don't need to store the entities_to_ignore.  Just move the mover along the path without any checks.  If the movement is ending, then we need to find a valid position.  Use `area_state::is_passable_size` with the mover's size for this, at that point in the process.

> 
> ```
> let mut anim = animation::move_animation::new(entity, path,
>                 entities_to_ignore.to_vec(),
>                 Config::animation_base_time_millis());
> ```
> 
> This enables the move_animation [`update()`](https://github.com/Grokmoo/sulis/blob/17b7b01e89651420f7c3651bc52b6ee2dde72815/sulis_state/src/animation/move_animation.rs#L85) function to check whether a position in the animation is occupied by a friendly unit.
> 
> ```
>     let p = model.path[frame_index];
>     let area_state = GameState::get_area_state(&model.area_id).unwrap();
>     let mut area_state = area_state.borrow_mut();
> 
>     if !area_state.move_entity(mover, p.x, p.y, move_ap as u32) {
>         marked_for_removal.set(true);
>         return;
>     }
> 
>     let index = (p.x + p.y * area_state.area.width) as usize;
> 
>     if area_state.entity_grid[index].iter().any(|i| model.entities_to_ignore.contains(i)) {
>         // do something here
>     } else {
>         // or here
>     }
> ```
> 
> We should
> 
>     1. try to predict that the remaining path is all on friend space, and then refuse to continue.

I've said this a few times already - we cannot predict ahead of time whether movement will be canceled or the mover will run out of AP.

We need to check *when the movement is ending* and place the mover if they are on an invalid square.

> 
>     2. walk on friend space, but store a walk_back for each on friend space, and add/execute the walk back (maybe in a bump-like animation) once we ended on a friend space. The walk_back should be cleared when we walk on private space.

We already store the `path` that the mover walked in the `move_animation`.  So, probably in the `cleanup` method we check if the mover position is invalid.  If it is, go back along the path until an open square is found, using `area_state::is_passable_size`.

It would be nice to make this walk back a smooth animation instead of a sudden jump on-screen.  But, I think that could be left for a second pass on this.

> 
> 
> We could also decide that, if the remainder of the path is all on friend space, then traps do not apply. I think this would allow for the remainder of the path to be accurately predicted [1], and would make the walk_back [2] storage unnecessary.

Doing this could cause campaign breakage in many places.  We rely on being able to trigger scripts when the player enters a given location.

**Comment by picobyte** [comment]
> We don't need to store the entities_to_ignore. Just move the mover along the path without any checks. If the movement is ending, then we need to find a valid position. Use `area_state::is_passable_size` with the mover's size for this, at that point in the process.

Ok, may have missed that.

> > We should
> > ```
> > 1. try to predict that the remaining path is all on friend space, and then refuse to continue.
> > ```
> 
> I've said this a few times already - we cannot predict ahead of time whether movement will be canceled or the mover will run out of AP.

I hope you don't mind that I'd like to understand your reasoning and discuss this. I understand that it's not possible in certain cases to predict the path, but aren't there cases where we can predict that we _can not finish, in any case,_ the path on a private space.

We've walked a few steps, and the terrain was rougher than expected; we got slowed horribly. We're running out of time. Now, we know we have less steps to go, but before me stand our friends. we know we're never make it past them in time. Will we tread their toes in a futile attempt to push our way through, knowing we will end up uncomfortably near to them, which they're never going to accept? They'd shove us right back, anyway. F*ck it, we can just wait right here for the next round. 

I may be wrong, but if not, this saves the walk_back, which is in my opinion the work-around solution.

> > We could also decide that, if the remainder of the path is all on friend space, then traps do not apply. I think this would allow for the remainder of the path to be accurately predicted [1], and would make the walk_back [2] storage unnecessary.
> 
> Doing this could cause campaign breakage in many places. We rely on being able to trigger scripts when the player enters a given location.

If the friend already triggered the trap, do we still step on the same stone? Or, if the trap is a chance thing, if a friend is already a on the spot and a trap goes off, do both receive the effects? Maybe you can give an example that illustrates that I'm wrong.

**Comment by picobyte** [comment]
I've adapted the cleanup function as suggested and now it seems to jump back as expected. There were some subtleties with borrowing, requiring a mover reference to go out of scope before we could call move_entity()
